### ISSUE TITLE
Consider `io_waiter::on_wait(std::optional<hal::callback>)`

### ISSUE THREAD

**Issue by kammce** [libhal/libhal#87]
When wait is called the callback is also called. With something like this, the IO_waiters can sequence a set of operations when wait is called. For example, if you are in a single threaded operation and want to kick off 3 DMA SPI transactions, this is a means to inject the 2nd and 3rd transfers into the original call. It requires additional stack space since the calls get deeper with each invocation of the sequence, but it should be relatively better in terms of memory utilization and performance than making additional threads.

I do understand that this does seem to approach the idea of a coroutines in a way except far less powerful. I believe that this kind of support is useful for situations where a developer doesn't want to invoke coroutines but still needs to do 3 actions in a row, such as a DMA SPI transfer of IMU data in parallel or commanding 6 smart motors to move in parallel without waiting for all of the motors to respond.

I'll had some code representing how this looks in the next comment.

**Comment by kammce** [comment]
This is something I wrote in the discord and pasting here:

# Tokenized callbacks for the exception analyzer

Writing this idea out here so I don't forget it. How to handle function pointers has been something on my mind for the exception analyzer. Now I have a simple cheap way of doing this.

So in libhal currently we may use something like `hal::callback<void(void)>` to create a callback to a function that returns nothing and takes no arguments. Nothing wrong with this in normal code. But for the exception analyzer, I've been thinking about how much of a headache this is. Basically, we will have no information within the binary to tell apart any function pointer of that signature with another. So calling any of those callback functions can emit any such exception that can be emitted from a `void(void)` function. If you only have a few thats not a big worry. If you have a TON of them, then you get chaos.

So I was thinking, is there anyway to split this up? Well a real easy way is to add a token to the signature. It doesn't do anything be change the function signature. 

So something like 

```C++
struct io_waiter_token {};
void set_callback(hal::callback<void(io_waiter_token)> p_callback);
```

The struct basically doesn't do anything but name another type. For the callback the token doesn't contain anything useful. But the signature is now different so the exception analyzer can see the groups of functions with this token as the signature and not just `void(void)`.

**Comment by kammce** [comment]
With that previous message, I'd like to make the API this:

```C++
// define a callback token for on_wait to differentiate it
struct io_waiter::on_wait_callback_t {}; 
io_waiter::on_wait(std::optional<hal::callback<void(on_wait_callback_t)>>);
```

The token is just there to breakup the signature compared to other types in the binary.

**Comment by kammce** [comment]
Using optional for callbacks for turning them on and off is actually quite ergonomic. See [godbolt](https://godbolt.org/z/jWj8hxjPa)

EDIT: updated the code from `on_wait_callback_t` to `on_wait_callback_tag`
